src/main/java/com/ncedu/cheetahtest/service/testscenario/TestScenarioServiceImpl.java

```diff
@@ -0,0 +1,94 @@
+        int totalTestScenarios = testScenarioDao.getTotalElements(idTestCase, title);
+        PaginationTestScenario paginationTestScenario = new PaginationTestScenario();
+        paginationTestScenario.setTotalElements(totalTestScenarios);
+        if (size * (page - 1) < totalTestScenarios) {
+            paginationTestScenario.setTestScenarios(testScenarioDao.findByTitleLike(title, idTestCase, size, size * (page - 1)));
+        }
+        return paginationTestScenario;
+    }
+
+    @Override
+    public TestScenario createTestScenario(TestScenario testScenario) {
+        TestScenario testScenarioWithSameTitle = testScenarioDao
+                .findTestScenarioByTitleExceptCurrent(
+                        testScenario.getTitle(),
+                        testScenario.getId());
+
+        TestCase testCase = testCaseGenDao.findById(testScenario.getIdTestCase());
+
+        if (testCase == null) {
+            throw new TestCaseNotFoundException();
+        }
+        if (testScenarioWithSameTitle != null) {
+            throw new TestScenarioAlreadyExistsException();
+        }
+
+        return testScenarioDao.createTestScenario(testScenario);
+    }
+
+
+    @Override
+    public PaginationItems getItemsFromScenario(int idTestScenario, int size, int page) {
+        int totalElements = testScenarioDao.getAllItemsAmount(idTestScenario);
+        PaginationItems paginationItems = new PaginationItems();
+        paginationItems.setTotalElements(totalElements);
+        if (size * (page - 1) < totalElements) {
+            paginationItems.setItemsFromTestScenario(testScenarioDao.getAllItems(idTestScenario, size, size * (page - 1)));
+        }
+        return paginationItems;
+    }
+
+    @Override
+    public PaginationTestScenario getAllTestScenariosFromTestCase(int idTestCase, int size, int page) {
+        int totalElements = testScenarioDao.getTestScenariosFromTestCaseAmount(idTestCase);
+        PaginationTestScenario paginationTestScenario = new PaginationTestScenario();
+        paginationTestScenario.setTotalElements(totalElements);
+        if (size * (page - 1) < totalElements) {
+            paginationTestScenario.setTestScenarios(testScenarioDao.findTestScenariosFromTestCase(idTestCase, size, size * (page - 1)));
+        }
+        return paginationTestScenario;
+    }
+
+    @Override
+    public TestScenario editTestScenario(TestScenario testScenario, int id) {
+        return testScenarioDao.editTestScenario(testScenario, id);
+    }
+
+    @Override
+    public PaginationTestScenario getAllTestScenarios(int size, int page) {
+        int totalElements = testScenarioDao.getAmountAllElements();
+        PaginationTestScenario paginationTestScenario = new PaginationTestScenario();
+        paginationTestScenario.setTotalElements(totalElements);
+        if (size * (page - 1) < totalElements) {
+            paginationTestScenario.setTestScenarios(testScenarioDao.getAllPaginated(size, size * (page - 1)));
```

Comment:
You should use here testCaseGenDao.getAllPaginated. Also if you use it, you won't have to calculate offset, just pass page and size.